Quick Shulker Support

kyrptonaught opened the pull request:
This pr adds support for Quick Shulker in the QuickShulkerHook class using the entrypoint provided by quickshulker. This pr also changes the inventory type in the container from SidedInventory to just Inventory allowing my custom inventories to be used. And adds a getter for slotCount in AbstractShulkerBoxBase

Note: 
- I had issues getting quickshulker added to the project using gradle, it will need to be manually added for now.
- This currently will crash on a dedicated server, inside QuickShulkerHook there is a call to a client side method ScreenProviderRegistry.INSTANCE.registerFactory

Commit:
Quick Shulker Support

This pr adds support for Quick Shulker container in the QuickShulkerHook class using the entrypoint provided by quickshulker. This pr also changes the inventory type in the container from SidedInventory to just Inventory allowing my custom inventories to be used.
Note: I had issues getting quickshulker added to the project using gradle, it will need to be manually added.
This currently will crash on a dedicated server, inside QuickShulkerHook there is a call to a client side only method ScreenProviderRegistry.INSTANCE.registerFactory

kyrptonaught commented:
Here is a copy of the -dev jar for quickshulker 
[quickshulker-1.1.0-dev.zip](https://github.com/i509VCB/BulkyShulkies/files/4238023/quickshulker-1.1.0-dev.zip)

i509VCB commented on gradle.properties:
```diff
@@ -13,6 +13,7 @@ fabric_version=0.4.29+build.290-1.15
 libcd_version=2.1.1+1.15.1
 cotton_resources_version=1.4.1+1.14.4
 shulkerboxtooltip_version=1.4.5+1.15
+quickshulker_version=-SNAPSHOT
```
Maybe try the commit for your version? Jitpack is weird.

kyrptonaught replied:
I tried multiple versions including using mavenLocal rather than jitpack. I’m assuming it’s an issue on quick shulkers end.

i509VCB replied:
Maybe try to setup something on bintray? I would like it to be possible to just be able to straight import the project without having to manually include your jar or make someone download it.

i509VCB commented on src/main/java/me/i509/fabric/bulkyshulkies/extension/QuickShulkerHook.java:
```diff
@@ -0,0 +1,81 @@
+      ItemStack stack = player.inventory.getInvStack(invSlot);
+      return  new ScrollableContainer(syncId, ShulkerBoxSlot::new, player.inventory, new ItemStackInventory(stack, getSlotCount(stack)), new TranslatableText("container.shulkerBox"));
+   }
+   public static int getSlotCount(ItemStack stack) {
```
I would reccomend instanceof checking the item before returning a slot count. If the item in the stack doesn't match a shulker box just return 0 or an OptionalInt

kyrptonaught replied:
Quick shulker should never give an item that isn’t registered, but it doesn’t hurt to be safe

i509VCB commented on src/main/java/me/i509/fabric/bulkyshulkies/api/block/base/AbstractShulkerBoxBlock.java:
```diff
@@ -279,7 +279,11 @@ public DyeColor getColor() {
+    	return this.slotCount;
+	}
+
+    public interface SingleTypePropertyRetriever<T> {
```
I use tabs for spacing in my code so you might want to change that.

kyrptonaught replied:
Will do

i509VCB replied:
This here still needs to be tabs, go set your IDE formatting to make a project code format for this and then make it defualt to tabs

i509VCB commented on src/main/java/me/i509/fabric/bulkyshulkies/extension/QuickShulkerHook.java:
```diff
@@ -0,0 +1,81 @@
+import me.i509.fabric.bulkyshulkies.client.screen.Generic13x7Screen;
+import me.i509.fabric.bulkyshulkies.client.screen.Generic9x7Screen;
+import me.i509.fabric.bulkyshulkies.client.screen.ScrollableScreen;
+import me.i509.fabric.bulkyshulkies.container.*;
```
Could you import all the contsiners individually? I don't use star imports.

i509VCB commented on src/main/java/me/i509/fabric/bulkyshulkies/extension/QuickShulkerHook.java:
```diff
@@ -0,0 +1,81 @@
+      ScreenProviderRegistry.INSTANCE.registerFactory(QS$SHULKER_9x7_CONTAINER, Generic9x7Screen::createScreen);
+      ScreenProviderRegistry.INSTANCE.registerFactory(QS$SHULKER_11x7_CONTAINER, Generic11x7Screen::createScreen);
+      ScreenProviderRegistry.INSTANCE.registerFactory(QS$SHULKER_13x7_CONTAINER, Generic13x7Screen::createScreen);
+      ScreenProviderRegistry.INSTANCE.registerFactory(QS$SHULKER_SCROLLABLE_CONTAINER, ScrollableScreen::createScreen);
```
I want this to work fine on servers, so possibly make a Client entrypoint. Also my screens have an Environment enchantment so this will fail.

i509VCB replied:
Also make sure this class uses tabs

kyrptonaught replied:
This will be addressed before release, not sure how I would like to approach this. Looking into alternatives

i509VCB replied:
One option is a Mixin Config plugin and a separate mixin which is only loaded on a client and when QuickShulker is present. If you do this please make sure you inject into your own QuickShulkerHook with `remap=false`

I have a trinkets example here (https://github.com/i509VCB/Project-F/blob/master/src/main/java/me/i509/fabric/projectf/mixin/compat/CompatMixinPlugin.java), but use a call to fabric loader first to verify we have a client and your mod is present.

i509VCB commented on src/main/java/me/i509/fabric/bulkyshulkies/extension/QuickShulkerHook.java:
```diff
@@ -0,0 +1,81 @@
+
+      QuickOpenableRegistry.register(((player, stack) -> ContainerProviderRegistry.INSTANCE.openContainer(QS$SHULKER_9x7_CONTAINER,player,
+              writer -> writer.writeInt(player.inventory.getSlotWithStack(stack)))),
+              ShulkerBlocks.GOLD_SHULKER_BOX);
```
Have you tested any of the colored boxes? Cause these only refer to the undyed blocks.

kyrptonaught replied:
They all do correctly work besides the ender slab as I just now realized they exist... whoops
Not too happen with this implementation  on quick shulkers end, but it avoids adding all 136 boxes individually

i509VCB replied:
There is a getItemStack method on BasicShulkerBox, which will give you any color of the same type (if you have a block instance). Just note you need to then copy the BlockEntityTag.

i509VCB reviewed (changes requested):
The logic seems to be place, just a few questions and style issues.

Please run a licenseFormat and make checkstyle happy. (gradle tasks for both)

i509VCB commented:
Also I have a little discord here where I can respond a bit quicker:
https://discord.gg/3zgUJse

Commit:
Updated to reflect changes in api,
Added Client side entrypoint to allow compatibility on servers,
Added instanceOf check when getting the slot count,
Added license

kyrptonaught commented:
Fixed some of your requested changes. Unable to run checkstyle due to not being able to get gradle to add quickshulker.

i509VCB commented on src/main/java/me/i509/fabric/bulkyshulkies/container/ScrollableContainer.java:
```diff
@@ -22,10 +22,12 @@
 
 import java.util.Arrays;
 
+import net.kyrptonaught.quickshulker.api.ItemStackInventory;
```
Remove unusued import here

i509VCB commented on src/main/java/me/i509/fabric/bulkyshulkies/container/GenericContainer11x7.java:
```diff
@@ -24,6 +24,7 @@
 
 package me.i509.fabric.bulkyshulkies.container;
 
+import net.kyrptonaught.quickshulker.api.ItemStackInventory;
```
Another unusued

i509VCB commented on src/main/java/me/i509/fabric/bulkyshulkies/client/extension/QuickShulkerHookClient.java:
```diff
@@ -0,0 +1,43 @@
+import net.fabricmc.fabric.api.client.screen.ScreenProviderRegistry;
+import net.kyrptonaught.quickshulker.api.RegisterQuickShulkerClient;
+
+public class QuickShulkerHookClient implements RegisterQuickShulkerClient {
```
This would work, but it's not refereed to by anything else.

kyrptonaught replied:
Forgot to push the changes to the fabric.mod.json, added another entrypoint

i509VCB reviewed (commented):
some more bits

i509VCB commented on src/main/java/me/i509/fabric/bulkyshulkies/extension/QuickShulkerHook.java:
```diff
@@ -0,0 +1,116 @@
+	public static ScrollableContainer createScrollable(int syncId, Identifier identifier, PlayerEntity player, PacketByteBuf buf) {
+		int invSlot = buf.readInt();
+		ItemStack stack = player.inventory.getInvStack(invSlot);
+		return new ScrollableContainer(syncId, ShulkerBoxSlot::new, player.inventory, new ItemStackInventory(stack, getSlotCount(stack)), new TranslatableText("container.shulkerBox"));
```
Usually we would want the text in the container to match the name of the block entity. If you could make a ShulkerTexts class in the registry package and refer to the text constants there it would work fine.

Commit:
Added client initializer that I forgot to push

kyrptonaught commented:
Well I also added ShulkerTexts and had that implemented making the containers have proper names. And added support for the ender slab. But my power flickered just enough to shut my pc off, conveniently after I hadn’t saved in a while. That’s nice 

And I also still can’t figure out this gradle issue, getting really frustrating

i509VCB commented:
Have you looked into putting quickshulker on Bintray instead?

kyrptonaught commented:
I don’t think the issue it jitpack, it does not work through mavenLocal either

Commit:
It works!
Added ShulkerTexts.java,
Removed unused imports
Added support for ender slab

Commit:
Fix crash on dedicated servers

i509VCB commented on build.gradle:
```diff
@@ -93,6 +97,7 @@ dependencies {
 
 	// Annotations
 	implementation group: 'org.jetbrains', name: 'annotations', version: "${project.jetbrains_annotations}"
+	implementation 'org.jetbrains:annotations:15.0'
```
Were the annotations not working when you added this?

i509VCB replied:
Okay so the missing texture box didn't work but I fixed that issue, I'll finish up with checkstyle and other issues and then merge it.

EDIT: I am going to add config options for whether these boxes can or can't be opened in hand.

i509VCB reviewed (commented):
Aside from that one issue, I am going to give the PR a test run and if it works I'll deal with fixing up the formatting

i509VCB commented:
Thanks for the PR, Manually merged: https://github.com/i509VCB/BulkyShulkies/commit/fb74a4cbc89bc6237acff28e89794b9870c447bf

i509VCB commented:
Will rebase onto 1.16.x branch in about 2 hours
